notebooks/rag/ragllama3/rag.py

```diff
@@ -0,0 +1,328 @@
+""" This is a RAG agent for MineDD"""
+
+# ---- standard python libraries ---- #
+import argparse
+import json
+import operator
+import os
+from os.path import exists
+from typing_extensions import TypedDict
+from typing import List, Annotated
+# ---- SQL ---- #
+import sqlite3
+# ---- langchain ----- #
+## - LLM
+from langchain_ollama import ChatOllama
+from langchain.schema import Document, AIMessage
+## - to chunk the text
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+## - to make/store embeddings
+from langchain_community.vectorstores import SKLearnVectorStore
+from langchain_community.embeddings.spacy_embeddings import SpacyEmbeddings
+from langchain_core.messages import HumanMessage, SystemMessage
+## - to build/run a langgraph
+from langgraph.graph import StateGraph, START, END
+
+local_llm="llama3.2:3b-instruct-fp16"
+db='../data/literature_relevant.db'
+embeddings_path=f"../embeddings/" + "union.bson"
+
+llm = ChatOllama(model=local_llm, temperature=0)
+llm_json_mode = ChatOllama(model=local_llm, temperature=0, format="json")
+database_path = db
+persist_path = embeddings_path
+
+# state class
+# this is the class state
+class GraphState(TypedDict):
+    """
+    Graph state is a dictionary that contains information we want to propagate to, and modify in, each graph node.
+    """
+    question : str # User question
+    generation : str # LLM generation
+    max_retries : int # Max number of retries for answer generation
+    answers : int # Number of answers generated
+    loop_step: Annotated[int, operator.add]
+    documents : List[str] # List of retrieved documents
+
+# Post-processing
+def format_docs(docs):
+    return "\n\n".join(doc.page_content for doc in docs)
+
+def extract_full_text_content(database_path):
+    conn = sqlite3.connect(database_path)
+    cursor = conn.cursor()
+    cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
+
+    text_content = []
+
+    # get fulltext table content
+    table_name = 'literature_fulltext'
+    cursor.execute(f"PRAGMA table_info({table_name});")
+    columns = cursor.fetchall()
+    column_types = [col[2] for col in columns]
+
+    # Retrieve all rows/papers from the table
+    cursor.execute(f"SELECT fulltext FROM {table_name};")
+    rows = cursor.fetchall()
+
+    # Iterate through the rows (which are papers) and extract text content
+    for row in rows:
+        row_text = []
+        for idx, value in enumerate(row):
+            if column_types[idx].lower() == 'text' and value is not None:
+                row_text.append(value)
+
+        if row_text:
+            text_content.extend(row_text)
+
+    conn.close()
+
+    return text_content
```

Comment:
```suggestion
    def extract_full_text_content(database_path):
    conn = sqlite3.connect(database_path)
    cursor = conn.cursor()
    cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")

    text_content = []

    # get fulltext table content
    table_name = 'literature_fulltext'
    cursor.execute(f"PRAGMA table_info({table_name});")
    columns = cursor.fetchall()
    column_types = [col[2] for col in columns]

    # Retrieve all rows/papers from the table
    cursor.execute(f"SELECT fulltext FROM {table_name};")
    rows = cursor.fetchall()

    # Iterate through the rows (which are papers) and extract text content
    for row in rows:
        row_text = []
        for idx, value in enumerate(row):
            if column_types[idx].lower() == 'text' and value is not None:
                row_text.append(value)

        if row_text:
            text_content.extend(row_text)

    conn.close()

    return text_content
```

suggestion:
```
def extract_full_text_content(database_path):
    conn = sqlite3.connect(database_path)
    cursor = conn.cursor()

    # Retrieve all rows/papers from the table
    cursor.execute(f"SELECT fulltext FROM literature_fulltext;")
    rows = cursor.fetchall()

    # Iterate through the rows (which are papers) and extract text content
    text_content = [row[0] for row in rows if isinstance(row[0], str) and row[0] is not None]

    conn.close()

    return text_content
```

this would be simpler and faster